src/main/scala/com/virtuslab/shared_indexes/remote/S3Upload.scala

```diff
@@ -0,0 +1,128 @@
+package com.virtuslab.shared_indexes.remote
+
+import com.intellij.indexing.shared.cdn.upload.{CdnUploadAnotherEntry, CdnUploadDataEntry, CdnUploadEntry}
+import com.intellij.indexing.shared.cdn.{CdnEntry, CdnUpdatePlan, S3, S3CdnEntry}
+import com.intellij.indexing.shared.local.LocalDiskEntry
+import org.slf4j.LoggerFactory
+import software.amazon.awssdk.core.sync.RequestBody
+import software.amazon.awssdk.services.s3.model.{DeleteObjectRequest, PutObjectRequest}
+
+import java.io.{ByteArrayInputStream, InputStream}
+import java.nio.file.Files
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future, TimeoutException}
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Success, Try, Using}
+
+class S3Upload(private val s3: S3, private val timeout: Duration = Duration(5, "min")) {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+  private val rootInBucket: String = s3.getRootInBucket
+  private implicit val ec: ExecutionContext = ExecutionContext.fromExecutor(s3.getDispatcher.getExecutor)
+
+  def updateS3Indexes(plan: CdnUpdatePlan): Unit = {
+    if (plan.getNewEntries.isEmpty && plan.getRemoveEntries.isEmpty) {
+      logger.warn("The update plan is empty, nothing to do!")
+      return
+    }
+
+    val failedRemoves = removeFiles(plan.getRemoveEntries.asScala)
+    val failedUploads = uploadFiles(plan.getNewEntries.asScala)
+    val failureSummary = new StringBuilder()
+    if (failedRemoves.nonEmpty) {
+      failureSummary.append(s"Failed to remove ${failedRemoves.size} entries: \n")
+      failedRemoves.foreach { case (entry, e) =>
+        failureSummary.append(s"  $entry: \n    ${e.getMessage}\n")
+      }
+    }
+    if (failedUploads.nonEmpty) {
+      failureSummary.append(s"Failed to upload ${failedUploads.size} entries: \n")
+      failedUploads.foreach { case (entry, e) =>
+        failureSummary.append(s"  $entry: \n    ${e.getMessage}\n")
+      }
+    }
+    if (failureSummary.nonEmpty) {
+      throw new RuntimeException(failureSummary.toString())
+    }
+  }
+
+  private def uploadFiles(files: Iterable[CdnUploadEntry]): Map[CdnUploadEntry, Throwable] = {
+    logger.info("Uploading {} items...", files.size)
+    val results = awaitAll(putObject)(files)
+    collectFailures(files, results)
+  }
+
+  private def removeFiles(files: Iterable[CdnEntry]): Map[CdnEntry, Throwable] = {
+    logger.info("Removing {} items...", files.size)
+    val results = awaitAll(deleteObject)(files)
+    collectFailures(files, results)
+  }
+
+  private def awaitAll[T, R](task: T => Future[R])(inputs: Iterable[T]): Iterable[Try[R]] = {
+    val futures = inputs.map(task)
+    val completions = futures.map(_.transform { case _: Try[R] => Success(()) })
```

Comment:
Interesting, why are we dumping failures here?

Reply:
This is to make sure that `Future.sequence` allows every future to complete, regardless of failures that may happen. Otherwise it would complete as soon as the first failure happens, without giving other futures a chance to complete.

The failures are actually collected in line 70.

Reply:
ah, right, you are using allCompleted to just wait, while `futures` still contains the original result